### ISSUE TITLE
Addition of permission templates for easier server admin

### ISSUE THREAD

**Issue by MajorTazer** [kraanzu/gupshup#1]
I've come from your advertisement on Reddit and think this project is very cool. However, I think the "power" concept is a bit confusing. As I understand it now, there is the `king` rank which has an infinite power level, thus allowing them to do any command. Any other rank can be assigned a numerical power level and this level changes what commands that rank can access.

The problem I've found is that the moderation commands such as `/ban`, `/mute`, etc. don't have a pre-defined power level - only the `king` rank can use them until the command is made accessible to a power level. This just seems unnecessary and overly complex to map each individual command to a power level.

Personally, I think it would be better if there were predefined power levels such as `1` for a moderator. Moderator commands should then be mapped to power levels of `1` or above. This provides a good foundation for permissions and the owner is able to take this and fine-tune it to their preferences. For example, if they wanted an `admin` rank with power level `2`, they could make moderators only be able to mute and admins be able to ban.

I've not had an in-depth look at this project, however, so if I've missed something, please let me know :)

EDIT: 
TL;DR IMO, there should be predefined power levels such as `1` for moderator and `2` for admin just to make setting up permissions easier since most owners are going to want roughly the same permission setup.

**Comment by kraanzu** [comment]
Yeah I get it... It would be way more simpler.. Since I wanted people to add any kind of rank names I thought this approach will be better :)
You can mention everything about the rank at the time of its creation,,, I forgot to add the power option... will add it too
like:
```
/add_rank <name> <icon> <color> <power_level>
/add_rank moderator 💀 red 1
```

**Comment by MajorTazer** [comment]
> Yeah I get it... It would be way more simpler.. Since I wanted people to add any kind of rank names I thought this approach will be better :)

I think the customisability of ranks and power levels is a good idea - just to make it easier, I think it'd be better if there were permission "templates" to make customisation easier. Most server owners would want roughly the same permission setup.

I don't mean that ranks should be limited to only `moderator`, `admin`, etc.

**Comment by kraanzu** [comment]
> I think it'd be better if there were permission "templates" to make customization easier

Yes I can do that..
Something like this for example?
```
/set_power_config discord
```
which will create ranks such as `A` and `B` with power levels `X` and `Y` automatically?

The `user/admin` can extend it by adding more ranks.. It will not be limited to just `A` and `B`.
Just a shortcut for writing a lot of commands

**Comment by MajorTazer** [comment]
> > I think it'd be better if there were permission "templates" to make customization easier
> 
> Yes I can do that.. Something like this for example?
> 
> ```
> /set_power_config discord
> ```
> 
> which will create ranks such as `A` and `B` with power levels `X` and `Y` automatically?

I saw that you plan on adding functionality for bots which would allow for a lot more commands to be available - therefore, it'd be a pretty cool idea to have third-party (made by the users) permission templates but for now, I think something like that would be good. Perhaps, something like:
```
/use_perm(ission)_template moderator
```
The only problem with the above is if the owner runs it with the moderator template and then another template, the moderator template may be overwritten so template compatibility would be something of a concern however, there aren't enough commands to have more than one "staff" role so I think moderator alone would be fine.

I assume there'd be functionality for the user to make their own templates? Probably in a JSON, YAML, etc. file format and it would be imported/exported.

Thoughts?

**Comment by kraanzu** [comment]
I plan to add functionality to upload files
We can use something like `/use_template discord.toml` which will upload the `discord.toml` file and will perform according to a specified format... report any errors like you mentioned:

> The only problem with the above is if the owner runs it with the moderator template and then another template, the moderator template may be overwritten


Is that what you meant by this?
> I assume there'd be functionality for the user to make their own templates? Probably in a JSON, YAML, etc. file format and it would be imported/exported.

**Comment by MajorTazer** [comment]
> > The only problem with the above is if the owner runs it with the moderator template and then another template, the moderator template may be overwritten
> 
> Is that what you meant by this?
> 
> > I assume there'd be functionality for the user to make their own templates? Probably in a JSON, YAML, etc. file format and it would be imported/exported.

There will always be conflict between third-party permission templates so there's not much you could do there. I was more on about conflict between in-built templates but if we only have one, that wouldn't be a problem.

**Comment by kraanzu** [comment]
Yeah we can make each one unique if possible :)

**Comment by MajorTazer** [comment]
> Yeah we can make each one unique if possible :)

Yep, or because they're in-built, you could add checks in each one but that all depends on if you're feeling lazy or not :P
We are programmers after all. 

If this is in development now, I'll close this issue.
